File: tools/ghidra_scripts/ExportDecompiledC.py
# ExportDecompiledC.py - Ghidra headless script
# Exports decompiled C for every function in the program.
# One file per ~1000 functions to keep file sizes manageable.
# Output directory set via WL2_OUTPUT_DIR environment variable.
#
# @category Warlords2

import os
from ghidra.app.decompiler import DecompInterface, DecompileOptions
from ghidra.util.task import ConsoleTaskMonitor

output_dir = os.environ.get("WL2_OUTPUT_DIR", "/tmp/wl2_ppc_decompiled")
if not os.path.exists(output_dir):
    os.makedirs(output_dir)

decomp = DecompInterface()
opts = DecompileOptions()
decomp.setOptions(opts)
decomp.openProgram(currentProgram)

monitor = ConsoleTaskMonitor()
functions = list(currentProgram.getFunctionManager().getFunctions(True))

CHUNK = 500
chunk_idx = 0
outfile = None
count = 0
errors = 0

for i, func in enumerate(functions):
    if i % CHUNK == 0:
        if outfile:
            outfile.close()
        chunk_idx = i // CHUNK
        fname = os.path.join(output_dir, "PPC_%04d.c" % chunk_idx)
        outfile = open(fname, "w")
        outfile.write("// Warlords II - Decompiled PPC Code\n")
        outfile.write("// Chunk %d (functions %d-%d)\n" % (chunk_idx, i, min(i+CHUNK-1, len(functions)-1)))
        outfile.write("// Decompiled by Ghidra %s\n\n" % getGhidraVersion())

    result = decomp.decompileFunction(func, 60, monitor)
    if result and result.decompiledFunction:
        src = result.decompiledFunction.getC()
        addr = func.getEntryPoint()
        outfile.write("// Function: %s at %s\n" % (func.getName(), addr))
        outfile.write("// Size: %d bytes\n" % func.getBody().getNumAddresses())
        outfile.write(src)
        outfile.write("\n\n")
        count += 1
    else:
        outfile.write("// Function: %s at %s -- decompile failed\n\n" % (func.getName(), func.getEntryPoint()))
        errors += 1

if outfile:
    outfile.close()

decomp.dispose()
print("Exported %d functions (%d errors) to %s" % (count, errors, output_dir))

File: tools/ghidra_scripts/ExportDecompiledC.java
// ExportDecompiledC.java - Ghidra headless script
// Exports decompiled C for every function in the program.
// One file per 500 functions. Output dir via WL2_OUTPUT_DIR env var.
//
// @category Warlords2

import ghidra.app.script.GhidraScript;
import ghidra.app.decompiler.DecompInterface;
import ghidra.app.decompiler.DecompileOptions;
import ghidra.app.decompiler.DecompileResults;
import ghidra.program.model.listing.*;
import ghidra.util.task.ConsoleTaskMonitor;
import java.io.*;
import java.util.*;

public class ExportDecompiledC extends GhidraScript {

    @Override
    public void run() throws Exception {
        String outputDir = System.getenv("WL2_OUTPUT_DIR");
        if (outputDir == null) outputDir = "/tmp/wl2_ppc_decompiled";

        new File(outputDir).mkdirs();

        DecompInterface decomp = new DecompInterface();
        DecompileOptions opts = new DecompileOptions();
        decomp.setOptions(opts);
        decomp.openProgram(currentProgram);

        FunctionIterator funcs = currentProgram.getFunctionManager().getFunctions(true);
        List<Function> funcList = new ArrayList<>();
        while (funcs.hasNext()) funcList.add(funcs.next());

        println("Exporting " + funcList.size() + " functions to " + outputDir);

        final int CHUNK = 500;
        PrintWriter out = null;
        int count = 0, errors = 0;

        for (int i = 0; i < funcList.size(); i++) {
            if (i % CHUNK == 0) {
                if (out != null) out.close();
                int chunk = i / CHUNK;
                String fname = outputDir + "/PPC_" + String.format("%04d", chunk) + ".c";
                out = new PrintWriter(new FileWriter(fname));
                out.println("// Warlords II - Decompiled PPC Code");
                out.println("// Chunk " + chunk + " (functions " + i + "-" + Math.min(i+CHUNK-1, funcList.size()-1) + ")");
                out.println("// Decompiled by Ghidra " + currentProgram.getMetadata().get("Created With Ghidra Version"));
                out.println();
            }

            Function func = funcList.get(i);
            DecompileResults result = decomp.decompileFunction(func, 60, new ConsoleTaskMonitor());
            out.println("// Function: " + func.getName() + " at " + func.getEntryPoint());
            out.println("// Size: " + func.getBody().getNumAddresses() + " bytes");
            if (result != null && result.decompileCompleted() && result.getDecompiledFunction() != null) {
                out.println(result.getDecompiledFunction().getC());
                count++;
            } else {
                out.println("// -- decompile failed --");
                errors++;
            }
            out.println();
        }

        if (out != null) out.close();
        decomp.dispose();
        println("Done: " + count + " ok, " + errors + " errors -> " + outputDir);
    }
}
